daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -212,11 +217,19 @@ private[lf] case class PartialTransaction(
     context: PartialTransaction.Context,
     aborted: Option[Tx.TransactionError],
     keys: Map[GlobalKey, Option[Value.ContractId]],
+    globalKeyInputs: Map[GlobalKey, Option[Value.ContractId]],
```

Comment:
_keys on the other hand tracks writes and reads. We could add some complicated logic to only rollback the "write" part of keys but I think that is much more confusing than just rolling back everything and adding a cache for the keys which is equivalent to only rolling back writes._


Is there an abstraction which would capture the relationship between these two elements of the state: ?
```
    keys: Map[GlobalKey, Option[Value.ContractId]],
    globalKeyInputs: Map[GlobalKey, Option[Value.ContractId]],
```
So it's easier to be sure that they are being properly kept in sync.

And perhaps a more evocative name for 'keys' which makes it clearer the relationship it has with `globalkeyinputs`?

(Maybe this is a refactoring for a followup PR.)

Reply:
Not really sure what a good abstraction would be but if you have any ideas I’m happy to change it although preferably in a separate PR.

Reply:
The code might be clearer if we use a custom scala type instead of `Option[Value.ContractId]`.
So we don't just have `None` to record the semantic info that a negative lookup has occurred.

Also the match statements we currently have like `case Some(Some(_)) =>` when we lookup in these maps would  be easier to read.

Reply:
good idea 3abd143265c527ad53658f075e496ca21f230dfa

Reply:
For a given key, I'm trying to understand which of the 9 states is possible, and what it means.
So far I have... Is this correct?

```
keys        globalKeyInputs

no-entry    no-entry    -- key never seem
no-entry    None        -- negative lookup; rolled-back
no-entry    Some(_)     -- positive lookup; rolled-back

None        no-entry    -- IMPOSSIBLE
None        None        -- negative lookup
None        Some(_)     -- ??

Some(_)     no-entry    -- IMPOSSIBLE
Some(_)     None        -- ??
Some(c1     Some(c2)    -- positive lookup -- (c1==c2)?
```

Reply:
(sorry, I see you already made a better type. But the question still stands)

Reply:
keys        globalKeyInputs

no-entry    no-entry    -- key never seen
no-entry    None        -- negative lookup; rolled-back -- EDIT: this could also be rolled back
no-entry    Some(_)     -- positive lookup; rolled-back

None        no-entry    -- IMPOSSIBLE -- Edit: Not quite true, just a global archive of a key that has not been looked up by key before will produce this. This is a bug and there is a TODO in the tests. https://github.com/digital-asset/daml/pull/9546 fixes this
None        None        -- negative lookup -- EDIT: this can also be a create & archive in the same transaction
None        Some(_)     -- ?? -- EDIT: lookup of a global contract & archive

Some(_)     no-entry    -- IMPOSSIBLE
Some(_)     None        -- ?? -- EDIT: just a create or negative lookup & then a create
Some(c1)    Some(c2)    -- positive lookup -- (c1==c2)? -- EDIT: c1==c2 iff we did not see an archive of c2